### DISCUSSION TITLE
Querying on children

### DISCUSSION THREAD

**Discussion by paradis-A** [appy-one/acebase#9] [Q&A] [answered]
let say i have this collection.
```javascript
//the collection sample
//parent collection here
stores = [
    {
        "some cuid": {
            "some props": "hello",
            // child collction
            products: [
                {
                    "some cuid": {
                        //some props
                    },
                },
            ],
        },
    },
]
//for example
db.ref("stores/some cuid").child('stores').push({
    // some props
})

```
how do i query products. i mean want to get all products.

**Comment by appy-one** [comment, marked as answer]
Hi, I'm not really sure what you are trying to do here?
First of all, you are mixing arrays and object collections. Only use arrays for small collections of data that do not change frequently, use object collections for all other situations.

If you want to maintain a collection of products, do this:
```js
db.ref('products').push({ title: 'product 1' }); // Add product to object collection
db.ref('products').push({ title: 'product 2' }); // Do it again
```
If you want all products you previously stored, simply get the value:
```js
const productsSnapshot = await db.ref('products').get();

// Iterate all objects in the collection with snapshot.forEach:
productsSnapshot.forEach(productSnapshot => {
   const product = productSnapshot.val(); // { "title": "Product 1" }
});

// Or, do it yourself by getting .val()
const products = productsSnapshot.val(); // { "somecuid1": { "title": "Product 1" }, "somecuid2": { "title": "Product 2" } }
```

If you want to query (instead of getting all), use ```.query```:
```js
const querySnapshots = await db.ref('products')
.query()
.filter('title', '>', 'Product')
.get()
```
See documentation for more info.

**Comment by paradis-A** [comment]
@appy-one Yay.  I missed some part of the documentation. So its not suggested to store array on nested object if it's frequently updated. I thought I can query from an array inside a nested object. just like I imagined.  I have to do the manual mappings after all. I should have put this to discussion tab maybe. Thanks for the fast reply anyway.

**Reply by appy-one** [reply]
Moved it to discussions

**Reply by appy-one** [reply]
It's ok to nest object collections, just don't use arrays for that purpose.
I initially misunderstood what you meant with "stores", probably because you did not provide any example properties. Here are some examples how to work with nested store/product collections:

```js
db.ref('stores/storeId/products').push({ title: 'product 1' }); // Add product to store's products object collection
db.ref('stores/storeId/products').push({ title: 'product 2' }); // Do it again
```

When creating a new store:
```js
const storeRef = await db.ref('stores').push({
   name: 'The Jeans Store',
   products: {} // Empty object collection
});
await storeRef.child('products').push({ title: 'Blue Jeans' }); // Create new product
```

To get a store's products:
```js
const productsSnapshot = await db.ref('stores/storeId/products').get();
```

**Comment by paradis-A** [comment]
Are you going to  implement after and before hooks on the acebase-server? just like on fastify

**Reply by appy-one** [reply]
What would you want to do with those hooks?

**Reply by paradis-A** [reply]
If it's not just me, i think its very convenient to have those. It'll allow us to decorate before or after entering. There are lots of thing we can do with those hooks. I.e. calling webhooks, custom, optional/extended validation, serverside fields filtering.

**Reply by appy-one** [reply]
You can add "cloud functions" to perform custom tasks upon data changes. Connect to your server with an AceBaseClient in a separate nodejs app, sign in as admin (or other user with rights to the data you want to read/write), add your custom event handlers.

```js
const db = new AceBaseClient({ host: 'localhost', port: 5757, dbname: 'mydb', https: false });
await db.ready();
await db.auth.signIn('admin', 'thepassword');

// Monitor images being added by users, resize them to multiple sizes:
db.ref('uploads/images').on('child_added', async snap => {
    const image = snap.val();
    const resizedImages = createImageSizes(image); // Some function that creates multiple image sizes
    const targetRef = await db.ref('images').push(resizedImages); // Store them somewhere else
    await snap.ref.remove(); // Remove original upload
});
```

To perform schema validation, add ".schema" nodes to your rules.json file on the server. See the [acebase-server documentation](https://www.npmjs.com/package/acebase-server).

Does this satisfy your needs? Is there something else you'd want to be able to do?

**Reply by appy-one** [reply]
I forgot to mention, another thing you can do on the server is create custom API methods with the ```server.extend(method, path, handler)``` method:

```javascript
const server = new AceBaseServer(...);
await server.ready();

const _quotes = [...];
server.extend('get', 'quotes/random', (req, res) => {
    let index = Math.round(Math.random() * _quotes.length);
    res.send(quotes[index]);
});
server.extend('get', 'quotes/specific', (req, res) => {
    let index = req.query.index;
    res.send(quotes[index]);
});
server.extend('post', 'quotes/add', (req, res) => {
    let quote = {
        text: req.body.text,
        author: req.body.author
    };
    quotes.push(quote);
    res.send('thanks!');
});
```

You can then call your API methods on the client side as follows:
```javascript
// Get random quote:
client.callExtension('get', 'quotes/random')
.then(quote => {
    console.log(`Got random quote: `, quote);
});

// Get specific quote:
client.callExtension('get', 'quotes/specific', { index: 15 })
.then(quote => {
    console.log(`Got quotes[15]: `, quote);
});

// Add a quote
client.callExtension('post', 'quotes/add', { 
    text: 'The greatest glory in living lies not in never falling, but in rising every time we fall', 
    author: 'Nelson Mandela' })
.then(result => {
    // "thanks!"
});
```

(Code above is from the [acebase-server documentation](https://www.npmjs.com/package/acebase-server))
